File: rust/src/solutions/intersection_of_two_arrays.rs
// LeetCode 349. Intersection of Two Arrays in Rust

use super::Solution;
use std::collections::HashSet;

impl Solution {
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut ans = Vec::new();
        let mut flag = HashSet::new();
        for num in nums1 {
            flag.insert(num);
        }
        for num in nums2.iter() {
            if flag.get(num).is_some() {
                ans.push(*num);
                flag.remove(num);
            }
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let nums1 = vec![1, 2, 2, 1];
        let nums2 = vec![2, 2];
        let ans = vec![2];
        let mut nums = Solution::intersection(nums1, nums2);
        nums.sort();
        assert_eq!(nums, ans);
    }

    #[test]
    fn example2() {
        let nums1 = vec![4, 9, 5];
        let nums2 = vec![9, 4, 9, 8, 4];
        let ans = vec![4, 9];
        let mut nums = Solution::intersection(nums1, nums2);
        nums.sort();
        assert_eq!(nums, ans);
    }
}

File: cpp/src/solutions/intersection_of_two_arrays.cpp
// LeetCode 349. Intersection of Two Arrays

#include "cpp/src/common/common.h"

class Solution {
 public:
  vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
    vector<int> ans;
    unordered_set<int> flag;
    for (auto num : nums1) {
      flag.insert(num);
    }
    for (auto num : nums2) {
      if (flag.find(num) != flag.end()) {
        ans.push_back(num);
        flag.erase(num);
      }
    }
    return ans;
  }
};

TEST(intersection, example1) {
  vector<int> nums1 = {1, 2, 2, 1};
  vector<int> nums2 = {2, 2};
  vector<int> ans = {2};
  auto nums = Solution().intersection(nums1, nums2);
  std::sort(nums.begin(), nums.end());
  EXPECT_EQ(nums, ans);
}

TEST(intersection, example2) {
  vector<int> nums1 = {4, 9, 5};
  vector<int> nums2 = {9, 4, 9, 8, 4};
  vector<int> ans = {4, 9};
  auto nums = Solution().intersection(nums1, nums2);
  std::sort(nums.begin(), nums.end());
  EXPECT_EQ(nums, ans);
}
